File: Leetcode/3507. Minimum Pair Removal to Sort Array I/solution.py
class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        def is_increasing_list(nums):
            if len(nums) == 1:
                return True
            for i in range(1, len(nums)):
                if nums[i] < nums[i-1]:
                    return False
            return True

        def reduce_min_sum_pair(nums):
            if len(nums) == 1:
                return nums
            current_min = inf
            min_index = 0
            for i in range(1, len(nums)):
                _sum = nums[i] + nums[i - 1]
                if _sum < current_min:
                    current_min = _sum
                    min_index = i
            nums[min_index - 1] = current_min
            del nums[min_index]
            return nums

        operations = 0
        while not is_increasing_list(nums):
            reduce_min_sum_pair(nums)
            operations += 1

        return operations

File: Leetcode/3507. Minimum Pair Removal to Sort Array I/solution.rs
impl Solution {
    pub fn minimum_pair_removal(mut nums: Vec<i32>) -> i32 {
        fn is_non_decreasing(nums: &Vec<i32>) -> bool {
            for i in 1..nums.len() {
                if nums[i] < nums[i - 1] {
                    return false;
                }
            }
            true
        }

        fn reduce_min_sum_pair(nums: &mut Vec<i32>) {
            let mut min_sum = i32::MAX;
            let mut min_index = 1;

            for i in 1..nums.len() {
                let sum = nums[i] + nums[i - 1];
                if sum < min_sum {
                    min_sum = sum;
                    min_index = i;
                }
            }

            nums[min_index - 1] = min_sum;
            nums.remove(min_index); // remove the right element of the merged pair
        }

        let mut operations = 0;

        while !is_non_decreasing(&nums) {
            reduce_min_sum_pair(&mut nums);
            operations += 1;
        }

        operations
    }
}